# Feature scaling with sklearn - Exercise

You are given a real estate dataset. 

Real estate is one of those examples that every regression course goes through as it is extremely easy to understand and there is a (almost always) certain causal relationship to be found.

The data is located in the file: 'real_estate_price_size_year.csv'. 

You are expected to create a multiple linear regression (similar to the one in the lecture), using the new data. This exercise is very similar to a previous one. This time, however, **please standardize the data**.

Apart from that, please:
-  Display the intercept and coefficient(s)
-  Find the R-squared and Adjusted R-squared
-  Compare the R-squared and the Adjusted R-squared
-  Compare the R-squared of this regression and the simple linear regression where only 'size' was used
-  Using the model make a prediction about an apartment with size 750 sq.ft. from 2009
-  Find the univariate (or multivariate if you wish - see the article) p-values of the two variables. What can you say about them?
-  Create a summary table with your findings

In this exercise, the dependent variable is 'price', while the independent variables are 'size' and 'year'.

Good luck!

## Import the relevant libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set() 

from sklearn.linear_model import LinearRegression 

## Load the data

In [3]:
data = pd.read_csv('real_estate_price_size_year.csv')

data.head()

price     size  year
0  234314.144   643.09  2015
1  228581.528   656.22  2009
2  281626.336   487.29  2018
3  401255.608  1504.75  2015
4  458674.256  1275.46  2009

In [4]:
#descriptive statistics
data.describe()

price         size         year
count     100.000000   100.000000   100.000000
mean   292289.470160   853.024200  2012.600000
std     77051.727525   297.941951     4.729021
min    154282.128000   479.750000  2006.000000
25%    234280.148000   643.330000  2009.000000
50%    280590.716000   696.405000  2015.000000
75%    335723.696000  1029.322500  2018.000000
max    500681.128000  1842.510000  2018.000000

## Create the regression

### Declare the dependent and the independent variables

In [5]:
x = data[['size','year']]

y = data['price']

### Scale the inputs

To scale the inputs we **NEED** to import **StandardScaler**

In [7]:
from sklearn.preprocessing import StandardScaler

In [8]:
#create our scaler object
scaler = StandardScaler()

In [9]:
#now fit the input data (features, inputs, aka x variable or independent)
scaler.fit(x)

StandardScaler()

In [10]:
#since preprocessing is finished, now we execute everything with transform
x_scaled = scaler.transform(x)

In [11]:
x_scaled

array([[-0.70816415,  0.51006137],
       [-0.66387316, -0.76509206],
       [-1.23371919,  1.14763808],
       [ 2.19844528,  0.51006137],
       [ 1.42498884, -0.76509206],
       [-0.937209  , -1.40266877],
       [-0.95171405,  0.51006137],
       [-0.78328682, -1.40266877],
       [-0.57603328,  1.14763808],
       [-0.53467702, -0.76509206],
       [ 0.69939906, -0.76509206],
       [ 3.33780001, -0.76509206],
       [-0.53467702,  0.51006137],
       [ 0.52699137,  1.14763808],
       [ 1.51100715, -1.40266877],
       [ 1.77668568, -1.40266877],
       [-0.54810263,  1.14763808],
       [-0.77276222, -1.40266877],
       [-0.58004747, -1.40266877],
       [ 0.58943055,  1.14763808],
       [-0.78365788,  0.51006137],
       [-1.02322731,  0.51006137],
       [ 1.19557293,  0.51006137],
       [-1.12884431,  0.51006137],
       [-1.10378093, -0.76509206],
       [ 0.84424715,  1.14763808],
       [-0.95171405,  1.14763808],
       [ 1.62279723,  0.51006137],
       [-0.58004747,

### Regression with SCALED FEATURES

In [12]:
#create regression object
reg = LinearRegression()

#Now we fit our standardized data into the regression
reg.fit(x_scaled,y)

LinearRegression()

### Find the intercept

In [13]:
reg.intercept_

292289.4701599997

### Find the coefficients

In [14]:
reg.coef_

array([67501.57614152, 13724.39708231])

### Calculate the R-squared

In [15]:
reg.score(x_scaled,y)

0.7764803683276793

### Calculate the Adjusted R-squared

$R^2_{adj.} = 1 - (1 - R^2)*\frac{n-1}{n-p-1}$

In [17]:
#get shape of x variale, which is now x_scaled

x_scaled.shape

(100, 2)

In [19]:
#identify n and p

#n is num of observation
n = x_scaled.shape[0]

#p is num of predictors 
p = x_scaled.shape[1]

r2 = reg.score(x_scaled,y)

In [20]:
#plug in variables into the adj R^2 formula

adjusted_r2 = 1 - (1-r2) * ((n-1)/(n-p-1))
adjusted_r2

0.77187171612825

### Compare the R-squared and the Adjusted R-squared

It seems the the R-squared is only slightly larger than the Adjusted R-squared, implying that we were not penalized a lot for the inclusion of 2 independent variables. 

### Compare the Adjusted R-squared with the R-squared of the simple linear regression

Comparing the Adjusted R-squared with the R-squared of the simple linear regression (when only 'size' was used - a couple of lectures ago), we realize that 'Year' is not bringing too much value to the result.

### Making predictions

Find the predicted price of an apartment that has a size of 750 sq.ft. from 2009.

In [21]:
reg.predict([[750,2009]])

array([78490785.31465861])

**THIS MAKES NO SENSE.**

**I THINK I NEED TO STANDARDIZE THE NEW DATA**

In [25]:
new_data = pd.DataFrame(data=[[750,2009]], columns=['size','price'])
new_data

size  price
0   750   2009

**REMEMBER,** scaler.transform is already primed up to do calculations as we set it up in the standardization process in the beggining. Proceeed below.

In [26]:
new_data_scaled = scaler.transform(new_data)

new_data_scaled

/Users/joeeey.d/opt/anaconda3/lib/python3.9/site-packages/sklearn/base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- price
Feature names seen at fit time, yet now missing:
- year

  warnings.warn(message, FutureWarning)


array([[-0.34752816, -0.76509206]])

In [27]:
#now our inputs are standardized. We can now predict.

reg.predict(new_data_scaled)

array([258330.34465995])

### Calculate the univariate p-values of the variables

For this I need to do **Feature Selection**. Key word is univariatie p-values!! When you see this word automatically think of Feature Selection

In [28]:
#import f_regression
from sklearn.feature_selection import f_regression

In [29]:
#add x and y variables into regression
f_regression(x_scaled,y)

(array([285.92105192,   0.85525799]), array([8.12763222e-31, 3.57340758e-01]))

**NOTE:** We have two arrays. The first array is the f-statistics and the second array is the p-values. The p-values are ordered in the way you added the inputs

In [30]:
p_values = f_regression(x_scaled,y)[1]
p_values

array([8.12763222e-31, 3.57340758e-01])

In [31]:
#lets round 
p_values.round(3)

array([0.   , 0.357])

**CONCLUSION:** Size has a p-value of 0.000 and Year has a p-value of 0.357

**Remember, if a variable has a P value above 0.05, we can disregard it!**

We can disregard year as its statistically insignificant

### Create a summary table with your findings

In [44]:
reg_summary = pd.DataFrame(data = x.columns.values, columns=['Features'])
reg_summary ['Coefficients'] = reg.coef_
reg_summary ['p-values'] = p_values.round(3)

In [45]:
reg_summary

Features  Coefficients  p-values
0     size  67501.576142     0.000
1     year  13724.397082     0.357